### ISSUE TITLE
fsspec filesystem handling link error

### ISSUE THREAD

**Issue by wolfv** [TheSnakePit/quetz#63]
with the recent changes the following error seems to happen on Fedora 32 and Ubuntu 16.04:

```
INFO:     127.0.0.1:38530 - "POST /api/channels/channel0/files/ HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/uvicorn/protocols/http/httptools_impl.py", line 385, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/fastapi/applications.py", line 181, in __call__
    await super().__call__(scope, receive, send)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/starlette/applications.py", line 102, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/starlette/middleware/sessions.py", line 75, in __call__
    await self.app(scope, receive, send_wrapper)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/starlette/routing.py", line 550, in __call__
    await route.handle(scope, receive, send)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/starlette/routing.py", line 227, in handle
    await self.app(scope, receive, send)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/starlette/routing.py", line 41, in app
    response = await func(request)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/fastapi/routing.py", line 196, in app
    raw_response = await run_endpoint_function(
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/fastapi/routing.py", line 149, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/starlette/concurrency.py", line 34, in run_in_threadpool
    return await loop.run_in_executor(None, func, *args)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/wolfv/Programs/quetz/quetz/main.py", line 400, in post_file
    handle_package_files(channel.name, files, dao, auth, force,
  File "/home/wolfv/Programs/quetz/quetz/main.py", line 458, in handle_package_files
    pkgstore.add_package(channel_name, file.file, dest)
  File "/home/wolfv/Programs/quetz/quetz/pkgstores.py", line 55, in add_package
    pkg.write(src.read())
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/fsspec/transaction.py", line 24, in __exit__
    self.complete(commit=exc_type is None)
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/fsspec/transaction.py", line 37, in complete
    f.commit()
  File "/home/wolfv/miniconda3/envs/quetz/lib/python3.8/site-packages/fsspec/implementations/local.py", line 244, in commit
    os.replace(self.temp, self.path)
OSError: [Errno 18] Invalid cross-device link: '/tmp/tmpfdh48fat' -> '/home/wolfv/Programs/quetz/test_quetz/channels/channel0/linux-64/coin-or-cgl-0.60.3-hf484d3e_0.tar.bz2'
```

I don't know yet how to fix it.

Any ideas @nowster ?

I am not sure why copyfileobj should attempt to link here...

**Comment by nowster** [comment]
That's because it appears to be using tempfile which uses TMPDIR environment variable which defaults to /tmp, and that's on a different filesystem to the destination, hence the rename can't work. Setting TMPDIR might help. Otherwise this may be a bug in `fsspec`.

The code I would have preferred to have used is (note the use of `mkstemp()`):
```
# Copyright 2020 Codethink Ltd
# Distributed under the terms of the Modified BSD License.

from contextlib import contextmanager
import os
import tempfile


@contextmanager
def AtomicOpen(filename, mode="w"):
    # Creates a tempfile in the same directory as the target filename.
    # Renames it into place when it's closed.
    fh, tmpname = tempfile.mkstemp(dir=os.path.dirname(filename), prefix=".")
    f = open(fh, mode)
    try:
        yield f
    except:  # noqa
        f.close()
        os.remove(tmpname)
        raise
    else:
        f.flush()  # Belt and braces (network file systems)
        f.close()
        os.rename(tmpname, filename)
```

**Comment by adriendelsalle** [comment]
Simply removing `with self.fs.transaction:` from both `LocalStore.add_file` and `LocalStore.add_package` seems to resolve the issue

**Comment by nowster** [comment]
Please reopen. The merged fix breaks index generation.

**Comment by adriendelsalle** [comment]
Sorry @nowster it was closed automatically by merging the workaround I proposed.

> The merged fix breaks index generation.

I understand the point but if you break capability to push packages, you then have nothing to index, which is not very useful.

Is the intent of this statement to enforce blocking file transactions (before any use such as indexing)?

For sanity, we should clearly introduce unit tests to avoid such breaking modifications.

I don't know how to test storage related to specific cloud providers and I'm wondering if such extensions should not be packaged outside the main/base package which provides an abstract class and basic implementation (as it was introduced with pkgstore and localstore), the same way [some Jupyterhub authenticators](https://github.com/jupyterhub/ldapauthenticator) are handled.
